src/views/SpikeLocations/SpikeLocationsWidget.tsx

```diff
@@ -0,0 +1,310 @@
+import { useSelectedElectrodes } from 'contexts/RecordingSelectionContext'
+import { useSelectedUnitIds } from 'contexts/UnitSelection/UnitSelectionContext'
+import BaseCanvas from 'FigurlCanvas/BaseCanvas'
+import { rectangularRegionsIntersect, RectangularRegion, transformPoint, Vec2, Vec4 } from 'FigurlCanvas/Geometry'
+import { useCallback, useMemo } from 'react'
+import { idToNum } from 'views/AverageWaveforms/AverageWaveformsView'
+import { computeElectrodeLocations } from 'views/AverageWaveforms/WaveformWidget/sharedDrawnComponents/electrodeGeometryLayout'
+import colorForUnitId from 'views/common/ColorHandling/colorForUnitId'
+import useDragSelectLayer from 'views/UnitLocations/useDragSelectLayer'
+import { AffineTransform, applyAffineTransform, applyAffineTransformInv, detAffineTransform } from 'views/UnitSimilarityMatrix/AffineTransform'
+import { useWheelZoom } from 'views/UnitSimilarityMatrix/MatrixWidget'
+import { defaultColors, ElectrodeColors } from '../AverageWaveforms/WaveformWidget/sharedDrawnComponents/electrodeGeometryPainting'
+
+export const defaultMaxPixelRadius = 25
+const circle = 2 * Math.PI
+
+export type Electrode = {
+    id: (number | string)
+    label: string
+    x: number
+    y: number
+}
+
+export type PixelSpaceElectrode = {
+    e: Electrode
+    pixelX: number
+    pixelY: number
+}
+
+export type LayoutMode = 'geom' | 'vertical'
+
+interface WidgetProps {
+    electrodes: Electrode[],
+    width: number
+    height: number
+    units: {
+        unitId: string | number
+        spikeTimesSec: number[]
+        xLocations: number[]
+        yLocations: number[]
+    }[]
+    colors?: ElectrodeColors
+    showLabels?: boolean
+    maxElectrodePixelRadius?: number
+    offsetLabels?: boolean
+    disableAutoRotate?: boolean
+    onlyShowSelected?: boolean
+}
+
+const defaultElectrodeLayerProps = {
+    showLabels: true,
+    maxElectrodePixelRadius: defaultMaxPixelRadius
+}
+
+
+const emptyDrawData = {}
+
+const markerRadius = 2
+
+const SpikeLocationsWidget = (props: WidgetProps) => {
+    const { width, height, electrodes, units, disableAutoRotate, onlyShowSelected } = props
+    const { selectedElectrodeIds } = useSelectedElectrodes()
+    const { selectedUnitIds, unitIdSelectionDispatch } = useSelectedUnitIds()
+    const {affineTransform, handleWheel} = useWheelZoom(width, height)
```

Comment:
I don't want to get too sidetracked here, and I haven't dug into this deeply yet--but how is this functionality separate from just tracking the native-space limits on the visible rectangle, maybe as part of the props to the locations widget? If we did that, we could additionally filter the data points that need to be displayed, and maybe reduce some redundant computation?

Reply:
This type of zoom is suitable when render is inexpensive (e.g., similarity matrix)